regression-tests/modules/dlso-gsql/dlsogsql.cc

```diff
@@ -0,0 +1,661 @@
+  rr.qtype = record->qtype;
+  auto qname = string(record->qname, record->qname_len);
+  rr.qname = DNSName(qname);
+  rr.qclass = QClass::IN;
+  rr.content = string(record->content, record->content_len);
+  rr.ttl = record->ttl;
+  rr.auth = record->auth;
+  rr.scopeMask = record->scope_mask;
+  rr.domain_id = record->domain_id;
+
+  auto dnsordername = DNSName();
+  if (ordername_ != nullptr) {
+    auto ordername = string(ordername_, ordername_len);
+    dnsordername = DNSName(ordername);
+  }
+  return handle->module->feedRecord(rr, dnsordername);
+}
+
+bool replace_record(
+  void* ptr,
+  uint32_t domain_id,
+  uint8_t qlen,
+  const char* qname,
+  uint16_t qtype,
+  uint16_t record_size,
+  const struct resource_record* records)
+{
+  auto* handle = static_cast<struct dlso_gsql*>(ptr);
+
+  vector<DNSResourceRecord> rrset;
+
+  for (uint16_t i = 0; i < record_size; i++) {
+    const struct resource_record* record = &(records[i]);
+
+    DNSResourceRecord rr;
+    rr.qtype = record->qtype;
+    string qname_ = string(record->qname, record->qname_len);
+    rr.qname = DNSName(qname_);
+    rr.qclass = QClass::IN;
+    rr.content = string(record->content, record->content_len);
+    rr.ttl = record->ttl;
+    rr.auth = record->auth;
+    rr.scopeMask = record->scope_mask;
+    rr.domain_id = record->domain_id;
+
+    rrset.push_back(rr);
+  }
+
+  auto qtype_ = QType(qtype);
+  auto qname_ = DNSName(string(qname, qlen));
+
+  return handle->module->replaceRRSet(domain_id, qname_, qtype_, rrset);
+}
+
+bool add_record_ent(void* ptr, uint32_t domain_id, bool auth, uint8_t qlen, const char* qname_)
+{
+  auto* handle = static_cast<struct dlso_gsql*>(ptr);
+
+  auto qname = DNSName(string(qname_, qlen));
+  map<DNSName, bool> nonterm;
+  nonterm.insert({qname, auth});
+
+  return handle->module->feedEnts(domain_id, nonterm);
+}
+
+bool add_record_ent_nsec3(void* ptr, uint32_t domain_id, uint8_t domain_len, const char* domain_, bool narrow, bool auth, uint8_t qlen, const char* qname_, const struct nsec3_param* ns3)
+{
+  auto* handle = static_cast<struct dlso_gsql*>(ptr);
+
+  auto qname = DNSName(string(qname_, qlen));
+  map<DNSName, bool> nonterm;
+  nonterm.insert({qname, auth});
+
+  DNSName domain;
+  if (domain_len > 0) {
+    domain = DNSName(string(domain_, domain_len));
+  }
+
+  NSEC3PARAMRecordContent ns3prc;
+  ns3prc.d_algorithm = ns3->alg;
+  ns3prc.d_flags = ns3->flags;
+  ns3prc.d_iterations = ns3->iterations;
+  ns3prc.d_salt = string(ns3->salt, ns3->salt_len);
+
+  return handle->module->feedEnts3(domain_id, domain, nonterm, ns3prc, narrow);
+}
+
+std::mutex g_configuration_mutex;
+
+extern "C" bool pdns_dlso_register(uint32_t abi_version, struct lib_so_api** ptr, bool dnssec, const char* args)
+{
+  auto* gsql = static_cast<struct dlso_gsql*>(malloc(sizeof(struct dlso_gsql)));
+  if (gsql == nullptr) {
+    return false;
+  }
+  auto* api = static_cast<struct lib_so_api*>(malloc(sizeof(struct lib_so_api)));
```

Comment:
I am quite rusty (sorry for the joke) with cpp, if there is a better way to write "I want owned memory and manage by myself the releasing" for this backend I would be happy to hear it.

Tried a little bit messing around with `unique_ptr` but wasn't exactly what I wanted since the release will occur on `api->release` call.

Also checking `abi_version` in test is pertinant will fix it tomorrow.

Reply:
How about wrapping `api` in a `unique_ptr` so you don't need to worry about freeing it in case of error, and if you want to transfer the ownership to `ptr`, and `ptr` cannot be a `unique_ptr` because of your ABI, do `*ptr = api.release()` just before returning?

Reply:
Same for `sql`, btw.